rcl/test/rcl/test_service.cpp

```diff
@@ -76,6 +77,30 @@ class CLASSNAME (TestServiceFixture, RMW_IMPLEMENTATION) : public ::testing::Tes
   }
 };
 
+void
+wait_for_server_to_be_available(
+  rcl_node_t * node,
+  rcl_client_t * client,
+  size_t max_tries,
+  int64_t period_ms,
+  bool & success)
+{
+  size_t iteration = 0;
+  bool is_ready = false;
+  rcl_ret_t ret = RCL_RET_OK;
+  do {
+    ++iteration;
+    ret = rcl_service_server_is_available(node, client, &is_ready);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    if (is_ready) {
+      success = true;
+      return;
+    }
+    std::this_thread::sleep_for(std::chrono::milliseconds(period_ms));
+  } while (iteration < max_tries);
+  success = false;
```

Comment:
@Barry-Xu-2018 consider returning this `bool` instead of `void` and a mutable reference for an output argument.

Reply:
Same below.

Reply:
@hidmic 

Thanks for your comments.  
If using **returning bool**, there is a macro expansion issue (Not decide to update the original logic **assert** code or using other ways).

In wait_for_server_to_be_available(), **assert** code is used.
```
  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;         ( contains a "return" statement inside)

    macro expanded:

  switch (0) case 0: default: if (const ::testing::AssertionResult gtest_ar = ( :: testing :: internal :: EqHelper < ( sizeof ( :: testing :: internal :: IsNullLiteralHelper ( 0 ) ) == 1 ) > :: Compare ( "0" , "ret" , 0 , ret ) )) ;
  else return ::testing::internal::AssertHelper( :: testing :: TestPartResult :: kFatalFailure , __FILE__ , 111 , gtest_ar . failure_message ( ) ) = ::testing::Message() << rcutils_get_error_string().str;
  

    build error:

/root/workspace/src/ros2/rcl/rcl/test/rcl/test_service.cpp: In function ‘bool wait_for_service_to_be_ready(rcl_service_t*, rcl_context_t*, size_t, int64_t)’:
/root/workspace/src/ros2/rcl/rcl/test/rcl/test_service.cpp:112:56: error: void value not ignored as it ought to be
   ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
                                                        ^~~
```

Reply:
I see. And yes, that happens because `ASSERT_*()` macros in `gtest` require the container function to return `void`. 

You can solve that by not using them there in the first place. I agree it's good to keep printing error strings, so maybe instead of a `bool`, you can return `rcl_ret_t` and do `ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str` on it.

Reply:
Thanks for your suggestion.